### ISSUE TITLE
deployments does not support (honor) containers restartPolicy 

### ISSUE THREAD

**Issue by ichekrygin** [kubernetes/kubernetes#24725]
Steps: Create deployment file and set restartPolicy to "Never"
Result: 
```
The Deployment "foo" is invalid.
spec.template.spec.restartPolicy: Unsupported value: "Never": supported values: Always
```

**Comment by AdoHe** [comment]
@kubernetes/kubectl we should document what restart policy deployment support.

**Comment by janetkuo** [comment]
@ichekrygin Deployments only support restartPolicy = `Always`; so do Replication Controllers, Replica Sets, and Daemon Sets. 

I'll document this.

**Comment by kachkaev** [comment]
Sorry for a late follow up, but what if I'd like to deliberately kill the pods instead of restarting? I've got a deployment with pods that are somewhat buggy. They do not restore themselves well after one of the containers fails and thus my deployment becomes unstable after some time. How can I maintain a Deployment (and consequently a ReplicaSet) that would kill and recreate failing Pods rather than restarting them?

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-apps-feature-requests

**Comment by kargakis** [comment]
@kachkaev https://github.com/kubernetes/kubernetes/issues/13488 is related to what you are asking.

**Comment by kachkaev** [comment]
Probably not quite. Rolling restart seems to be something triggered by a user, while what i’m asking is pod recreation on failure (rather than a restart).

**Comment by yuedai** [comment]
I have a similar requirement. @kachkaev

**Comment by mattmb** [comment]
Does anyone know *why* deployments only support restartPolicy == Always? It's not clear from the documentation and I can't spot the reason easily in the code.

I think there are legitimate use cases for "Never". For example, I have a pod running on a node and it has a liveness probe. That node then fails in someway that makes it remain up but perform slowly. The liveness probe fails and the kubelet will now try to recreate the container. But the problem is with the node so the new container doesn't become healthy. And my ReplicaSet doesn't see the pod as failed so it doesn't try to create a new pod elsewhere.

If I could set the restart policy to Never then I imagine in this scenario: liveness probe fails, container terminated, pod marked as failed, replica set creates a new pod (maybe or maybe not on the same node).

Or have I mis-understood something?

**Comment by kyessenov** [comment]
Another example that is difficult to handle: sidecars and sidecar proxies, in particular. If an application container does not properly clean up after itself (e.g. UDS sockets), then the pod stays alive since the sidecar is still running. The application keeps trying to restart, and keeps failing since it attempts to use some resources that have not been cleaned up.

**Comment by llech** [comment]
If deploymets support only restartPolicy: Always, why do that parameter exists at all?
It doesn't make much sense to have a 'parameter' that can have only one value...

What is the alternative for running one-time or periodic tasks, like data import or backups?

**Comment by rcsavage** [comment]
> If deploymets support only restartPolicy: Always, why do that parameter exists at all?
> It doesn't make much sense to have a 'parameter' that can have only one value...
> 
> What is the alternative for running one-time or periodic tasks, like data import or backups?

I totally agree, there's no need for the parameter if it's behaves like this.

**Comment by chrisghill** [comment]
I also have a use case like @kachkaev where I would prefer to have a pod die and be rescheduled as a new pod than be restarted.  This feature would be useful.

**Comment by kaneg** [comment]
I was also confused by the failed message which said 'Unsupported value: "Never": supported values: "Always"'.
There is no any document saying that the flag can't be supported by Deployment.

**Comment by 2PacIsAlive** [comment]
This has me so confused. I am creating a deployment that automatically creates a replicaset - I am not creating the replicaset directly. How do I get that replicaset to have `restartPolicy` `Never`?

**Comment by symonsoft** [comment]
I'm experiencing the same problem. Uh three years since ticked opened. I think the worst decision in my life is the decision to use fkn Kubernetes at all. The worst product ever.

**Comment by SummerShoohaw** [comment]
I'm experiencing the same problem +1, we have some sidecars running in our pod which will keep running forever, and the we want the pod to fail and restart if the primary container failed.

**Comment by janpaul123** [comment]
Sounds like this should get addressed by https://github.com/kubernetes/enhancements/pull/912. My workaround for now has been to set up a ServiceAccount that can delete pods, and then run `kubectl delete pods $HOSTNAME` from within my container.

**Comment by wuestkamp** [comment]
I would also like to completely start a new pod instead restarting. @janpaul123 your workaround sounds ugly as hell, but thanks for suggestion ;)

**Comment by AKCen** [comment]
I wonder this thing was raised 3 years back and why is documentation here https://v1-12.docs.kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy says 

```
A PodSpec has a restartPolicy field with possible values Always, OnFailure, and Never. 
```

while I get 

```
... is invalid: spec.template.spec.restartPolicy: Unsupported value: "Never": supported values: "Always"
```

kubectl version

```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.8", GitCommit:"a89f8c11a5f4f132503edbc4918c98518fd504e3", GitTreeState:"clean", BuildDate:"2019-04-23T04:41:47Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by 0marq** [comment]
> Only a .spec.template.spec.restartPolicy equal to Always is allowed, which is the default if not specified.

The doc says only Always is supported for Deployments

You might want to consider switching to a Job 

> Only a RestartPolicy equal to Never or OnFailure is allowed.

See 
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template
https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#pod-template

**Comment by felix0080** [comment]
why closed,this issue is important

**Comment by wuestkamp** [comment]
I also think it would be great to reopen this.

**Comment by z-oo** [comment]
Please reopen because this is critical and particularly bad when combined with unstable volumne mount bug like this: https://github.com/kubernetes/kubernetes/issues/67643 or a dead fuse daemon.

If any volumne mount is broken for a pod, we have no choice but restarting the whole pod. We have put liveness probe and in code health check to commit suicide after write failure or any mountpoint abnormality. But if restartPolicy is always, the pod simply enters crashloop without restarting itself.

We used to be able to set "Never" on deployments and it works as expected. However recently we found that this has been enforced and our existing mechanism stops working.

Another possible solution is to add a time limit of crashloopbackoff and kill the pod after limit. We can write a script to do so but it would be better if it can be built into the official controller or kubelet.

**Comment by yurilaaziz** [comment]
hope this comes true. I would like to avoid seeing failed deployment with 200x pod restarts,
Okay to keep restartPolicy to Always as default.
but forcing Never will help.

**Comment by tommyjcarpenter** [comment]
why is this issue closed? how do I get k8s to not restart a pod? what is the point of a parameter if it can only take one value?

**Comment by Nitesh-vaidyanath** [comment]
+1 Our application is designed in such a way if container crash it will fail to start again.  We need option "spec.template.spec.restartPolicy" : "Never".

**Comment by sinchanroy** [comment]
+1, our pods gets restarted 200x in case of errors and eating up much of resources. I would voteup for this issue

**Comment by junneyang** [comment]
+1 Our application is designed in such a way if container crash it will start again, if not we want to just ignore the health check error,  We need option "spec.template.spec.restartPolicy" : "OnFailure".

**Comment by bronger** [comment]
My backup pod contained a bug and restarted over and over again, wearing out the backup disk. This is not good! And any remedy is more complicated than being able to set restartPolicy=Never.

**Comment by fara-nak** [comment]
Please open this, we need the pod not to restart, what's the point of "Never" if we can't use it?

**Comment by bronger** [comment]
Well, to be fair, the parameter exists because it is inherited from pod spec. A deployment spec contains a pod spec with some constrains, one of them being that restartPolicy can only be Never.

I *think* the reasoning behind this is that deployments are not jobs.  Like a container must be able to cope with the fact that it can be shut down at any time, a deployment must be able to cope with the fact that it is kept alive by restarts.

**Comment by huide9** [comment]
+1 need it can be configured as onFailure or Never.  Always is not fit for the real case here.

**Comment by amboowang** [comment]
+1 need it can be configured as onFailure or Never. Always is not fit for the real case here.

**Comment by NovaHe** [comment]
+1 need it can be configured as onFailure or Never. Always is not fit for the real case here.

**Comment by persunde** [comment]
Please add onFailure and Never.
+1 need it can be configured as onFailure or Never. Always is not fit for the real case here.

**Comment by bronger** [comment]
Please don’t spam.

**Comment by viniciusd** [comment]
> Only a RestartPolicy equal to Never or OnFailure is allowed.



> > Only a .spec.template.spec.restartPolicy equal to Always is allowed, which is the default if not specified.
> 
> The doc says only Always is supported for Deployments
> 
> You might want to consider switching to a Job
> 
> > Only a RestartPolicy equal to Never or OnFailure is allowed.
> 
> See
> https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template
> https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#pod-template

Jobs are immutable, I need a run-to-completion that can be shot again and again whenever I want.

**Comment by rainerfrey** [comment]
Comments in this issue mix to separate aspects:
* run-to-completion or fail-and-exit use cases (these are addressed by Jobs)
* use cases that want to keep the application running, but require the restart of the entire pod instead of just the container
The second case doesn't seem to be addressed in any way yet. Another example for that is: a main container, that needs the action of an init container to be run on every start.

**Comment by viniciusd** [comment]
>     * run-to-completion or fail-and-exit use cases (these are addressed by Jobs)
> 
The issue with jobs for me is the fact one can't run a job with the same name. Even if I use a plugin that adds random gibberish to the job name, they are going to be pilling up in the cluster. Lots of pollution after some time.

**Comment by wuestkamp** [comment]
We did need the use case: `want to keep the application running, but require the restart of the entire pod instead of just the container`.

For this we let our main container, after its done, contact the K8s api to delete its own Pod (without gracePeriod) which causes a new Pod to be created.

Example here, but stripped down so not sure if working just like that:



```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: agent
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: agent
  namespace: default
rules:
  - apiGroups:
      - ""
    resources:
      - pods
    verbs:
      - delete
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: agent
  namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: agent
subjects:
  - kind: ServiceAccount
    name: agent
    namespace: default
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: agent
  namespace: default
spec:
  replicas: 10
  selector:
    matchLabels:
      id: agent
  template:
    metadata:
      labels:
        id: agent
    spec:
      serviceAccountName: agent
      containers:
      - name: sidecar
        image: sidecar
        command: ["./sidecar"]
      - name: agent
        image: agent
        command:
        - /bin/sh
        - -c
        - >
          ./agent; # processes one message then exits
          curl -m 10 -k -X DELETE --data '{"gracePeriodSeconds":0,"propagationPolicy":"Background"}' -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)" -H 'Accept: application/json' -H'Content-Type: application/json' https://kubernetes.default/api/v1/namespaces/$(cat /run/secrets/kubernetes.io/serviceaccount/namespace)/pods/$(hostname);
          sleep 10 # awaiting death while sleeping

```

**Comment by scarby** [comment]
> > ```
> > * run-to-completion or fail-and-exit use cases (these are addressed by Jobs)
> > ```
> 
> The issue with jobs for me is the fact one can't run a job with the same name. Even if I use a plugin that adds random gibberish to the job name, they are going to be pilling up in the cluster. Lots of pollution after some time.

If your expectation is that your containers exit you should use a job. If you are concerned about these building up then use a label and delete the old jobs before you spin up a new one. Also time stamping is better than random gibberish.

**Comment by amendoza-navent** [comment]
Is there a workaround for this? I'm testing an application to see a failure cause, and the constant restarts make noise in logs.

**Comment by diranged** [comment]
I feel like a broken record.. but this is a really pretty big issue. There are _tons_ of cases where you do not want to restart a broken pod, but replace it. I am really surprised at how little movement there is here. :/

**Comment by 2PacIsAlive** [comment]
> I feel like a broken record.. but this is a really pretty big issue. There are _tons_ of cases where you do not want to restart a broken pod, but replace it. I am really surprised at how little movement there is here. :/

Just use a pod then, you don't need a deployment

**Comment by diranged** [comment]
> > I feel like a broken record.. but this is a really pretty big issue. There are _tons_ of cases where you do not want to restart a broken pod, but replace it. I am really surprised at how little movement there is here. :/
> 
> Just use a pod then, you don't need a deployment

That is not what we are saying.. we are saying that there are deployment-like models where you cannot or do not want the pod to restart in a failure, you want a replacement pod that has a new identity.

**Comment by bronger** [comment]
It’s part of the nature of the deployment that it keeps things alive. If this is unfortunate for you, then deployment is not what you should use.

In interesting use case mentioned here that some people need to tear down and recreate the whole pod. @diranged You may want that, too? But note that this could not be achieved with restartPolicy anyway. It would need a wholly new top-level setting in the YAML spec.

**Comment by diranged** [comment]
> It’s part of the nature of the deployment that it keeps things alive. If this is unfortunate for you, then deployment is not what you should use.
> 
> An interesting use case mentioned here is that some people need to tear down and recreate the whole pod. @diranged You may want that, too? But note that this could not be achieved with restartPolicy anyway. It would need a wholly new top-level setting in the YAML spec.

Ok, so I think its clear from the community that there is a desire to have a way of saying "If Pod-XYZABC exits for any reason, do not _restart_ it. Let it die and go away." I think beyond that, the mechanics of how we accomplish such a thing is up to the experts in the community who understand the tooling the most.

As a _non_ expert, my mental model is that a `Deployment` handles replacing Pods that go away quite well, so if we could just tell the PodSpec to "go away when the processes exit", that _feels_ like the answer I think people are grasping for. However, perhaps that is simply not how the interaction between a Pod and Deployment and the Scheduler works... so maybe the parameter should be somewhere else.

I will note though that its clear the community is confused by the appearance of the `restartPolicy` parameter on the PodSpec itself in this case (though, the argument that it's just _part_ of the PodSpec, so it can't be hidden is reasonable). It just seems to me like the most intuitive thing is for the behavior to be the following:

  * DeploymentA creates PodA, PodB, PodC (with `restartPolicy: Always`)
  * PodC exits (either 0 or 1 exit code)
  * PodC is restarted by the Kubelet, and the Deployment is none the wiser..

Alternatively

  * DeploymentA creates PodA, PodB, and PodC with `restartPolicy: Never`
  * PodC exits (either 0/1 exit code)
  * PodC is then terminated by the Kubelet
  * DeploymentA then creates PodD to replace the PodC which went away

**Comment by tommyjcarpenter** [comment]
Yes; honestly that was the most frustrating part of this issue for me. If I can't change it, remove it as a parameter. Having a parameter that can only be set to a single value is a frustrating UX experience. It's like that South Park episode where Cartman buys an amusement park but no one is allowed to buy tickets to enter.

**Comment by bronger** [comment]
> Yes; honestly that was the most frustrating part of this issue for me. If I can't change it, remove it as a parameter. Having a parameter that can only be set to a single value is a frustrating UX experience. It's like that South Park episode where Cartman buys an amusement park but no one is allowed to buy tickets to enter.

Deployment templates are pod specs, and this is good because you specify a pod after all. And a pod may have an arbitrary restartPolicy. But as part of a deployment spec, there is the additional constraint that restartPolicy can only be “always”.

The alternative would be to have pod-spec and pod-within-deployment-template-spec, which I would find even more confusing.

**Comment by tommyjcarpenter** [comment]
I understand how it became to be - but that does not make it less frustrating (as evidenced by this thread, which would be 1/10th  the size if the parameter was not exposed at all)

**Comment by bronger** [comment]
Well, frankly, such a thread is hardly representative.

FWIW, I prefer a documentation that says “A is [long spec], and B is like A, but with some little changes” instead of “A is [long spec], and B is [long spec]”.

**Comment by bronger** [comment]
But back on topic, I want to point out that restartPolicy refers to single containers, not to the pod as a whole. This, the use case that you need to replace the whole pod (as mentioned by https://github.com/kubernetes/kubernetes/issues/24725#issuecomment-365758473 and I think also by https://github.com/kubernetes/kubernetes/issues/24725#issuecomment-825047795) needs to be addressed in a different way anyway. restartPolicy cannot help here.

**Comment by cwrau** [comment]
We would like this functionality too, particularly `restartPolicy: OnFailure`.

Our application sometimes needs to restart, but that triggers alerts because the Pod restarted. Currently, there is no metric `failureRestart` vs `successRestart` so we can't distinguish between these two cases.

If the Pod just got deleted on container exit and then recreated that would solve that little annoyance for us.

But our maybe weird use-case aside, one big question which I didn't find answered here is still open;

*Why* does this constraint exist in the first place?

Maybe if you could explain why the DeploymentSpec only allows PodSpecs with `restartPolicy: Always` we can accept that, but as it stands now I can only surmise that it's just some arbitrary decision that has been made

**Comment by scarby** [comment]
> Maybe if you could explain why the DeploymentSpec only allows PodSpecs with restartPolicy: Always we can accept that, but as it stands now I can only surmise that it's just some arbitrary decision that has been made

@cwrau This has been explained in this thread. The deployment controller is designed for maintaining availability of long running services that are not routinely designed to exit. (Web/application servers, queue workers that poll for new work forever etc.)

Having a deployment with a restart policy of OnFailure is basically akin to having a Job these are designed to run to completion and not restart

Having a restart policy of Never and you may as well directly create the pods.

I'd suggest that if you need some other behavioral it's a different class of thing and you should consider writing a custom controller to handle that.

**Comment by cwrau** [comment]
Mh, I must've missed that, sorry

I kinda understand that, but I don't see a difference between restarting a Pod and just letting it die and creating a new one; the end result is the desired number of replicas

The same thing that happens when you delete a Pod or a node dies; the current number of replicas is below the desired one, therefore the ReplicaSet controller creates a new pod

Additionally, it doesn't click for me why it's ***necessary*** for Deployments to be `restartPolicy: Always`

I get that that's your desired concept, but I don't see any drawbacks to allowing other policies and leaving `Always` as the default and on the other side I see quite a few advantages, like that this portion of the community would benefit from this.

Writing a custom controller is not something you'd write in a day and push to production, but I'd think allowing other policies would be.

As for your examples;

I wouldn't say that it's in any way close to being akin to a Job, as Jobs run to completion and stop. Deployments keep a number of replicas up, forever.

Also, there is a very big difference between a Deployment with `restartPolicy: Never` and creating the Pods manually; the first one is automatic.

There are also advantages like https://github.com/kubernetes/kubernetes/issues/24725#issuecomment-561396435;

> Please reopen because this is critical and particularly bad when combined with unstable volumne mount bug like this: #67643 or a dead fuse daemon.
> 
> If any volumne mount is broken for a pod, we have no choice but restarting the whole pod. We have put liveness probe and in code health check to commit suicide after write failure or any mountpoint abnormality. But if restartPolicy is always, the pod simply enters crashloop without restarting itself.
> 
> We used to be able to set "Never" on deployments and it works as expected. However recently we found that this has been enforced and our existing mechanism stops working.

**Comment by rainerfrey** [comment]
> I kinda understand that, but I don't see a difference between restarting a Pod and just letting it die and creating a new one; the end result is the desired number of replicas
> 
> The same thing that happens when you delete a Pod or a node dies; the current number of replicas is below the desired one, therefore the ReplicaSet controller creates a new pod 
> As for your examples;
> 
> I wouldn't say that it's in any way close to being akin to a Job, as Jobs run to completion and stop. Deployments keep a number of replicas up, forever.
> 
> Also, there is a very big difference between a Deployment with `restartPolicy: Never` and creating the Pods manually; the first one is automatic.

I completely agree, considering this would be very much appreciated.

**Comment by bronger** [comment]
I still have the impression that different use cases are mixed up here. Just so that I understand what is discussed at the moment: You need a setting “restartTogether” orthogonal to “restartPolicy” that says that *if* at least one _container_ dies, all containers in that pod should be terminated and restarted?

And if that restartTogether existed, you could live with restartPolicy=always?

**Comment by pre** [comment]
I think `restartPolicy: recreate` is more like it. It’s not about restarting, it’s more of ”terminate this pod and replace it with a new one” instead of trying to restart containers.

**Comment by bronger** [comment]
The life cycles of container and pod are very different in k8s. Killing the pod is different from terminating all of its containers. If you really kill the pod, it needs to be re-scheduled. This may hang. The pod consumes the node’s resources for having space for your containers. Killing the pod means losing this reservation.

I don’t think this is a good idea. Besides, why is restarting all containers within the pod _in situ_ not sufficient?

**Comment by cwrau** [comment]
Good point, I hadn't thought about that

I guess at least our use case, *not* getting alerts when a container actively wants to restart by exiting with `0`, could just be solved with a different label on the `kube_pod_container_status_restarts_total` metric.

But some use cases may have to take that risk, like the one I mentioned above; https://github.com/kubernetes/kubernetes/issues/24725#issuecomment-561396435

**Comment by bronger** [comment]
https://github.com/kubernetes/kubernetes/issues/24725#issuecomment-561396435 should get away with restarting the containers, leaving the pod intact, shouldn’t it?

**Comment by cwrau** [comment]
I don't know myself, but it *sounds* like it doesn't, maybe @z-oo can shed some more light on this?

**Comment by rainerfrey** [comment]
> I still have the impression that different use cases are mixed up here. Just so that I understand what is discussed at the moment: You need a setting “restartTogether” orthogonal to “restartPolicy” that says that _if_ at least one _container_ in a _pod_ dies, all containers in that pod should be terminated and be restarted? (The pod remains existing, in particular, its scheduling, its name, its connection with persistent volume claims etc.)
> 
> And if that restartTogether existed, you could live with restartPolicy=always?

I guess you don’t address me specifically, but from my perspective, in the case that brought me to this issue, I’d say yes that would solve it, *if* restartTogether includes executing init containers.

**Comment by bronger** [comment]
> […]
> 
> In general most of the above is fixed if "restartPolicy: Never" is accepted, and new pods are created on failure, or as mentioned a new type of "ReplicaSet" is created to help this behaviour.

The “new pods are created on failure” would indeed solve some of the mentioned issues. However, I am not so sure about the “restartPolicy: Never”, which is the actual topic of this issue. It would simply stop K8s from doing anything further, and a human or external tool would have to intervene. I know that people in this issue have seldomly suggested “restartPolicy: Never” in an isolated manner, still:

Honestly, the thread has shifted and diverged its subject enough so that it should be continued in new issues with clearly defined use cases and suggestions.

**Comment by PicoCreator** [comment]
Hmm i think there is confusion, cause i think some users interpration of "restartPolicy: Never" + gurantee X replica, would mean that new pods would be created to meet the replica requirement. 

But no worries, less concerned about the actual language of the fearture, and more concerned of it being supported (like restartPolicy: Recreate???), etc.

Created as a seperate issue: https://github.com/kubernetes/kubernetes/issues/101933

**Comment by devzer01** [comment]
The Never value is required for some troubleshooting purposes so that logs are not lost during development stages when time crunching matters like fluentd configuration is not complete. Kindly add the restart policy Never support

**Comment by lordnynex** [comment]
Just my 2 cents, I need restart never to run load testing tools. I can not launch as a pod (which supports restart never) and have > 1 replicas. Naturally I shifted to looking at a deployment but the same incompatibility exists. If my pods restart it will result in a never ending load test. In the case of my database load testing tools, they end up sampling rows out of a database to generate read operations. Over time that sample set gets larger and larger because the pods are restarting. This effectively halts the cluster. 

I am looking forward to ephemeral containers for this use case, but I fear they may have the same drawbacks. I need a way to launch containers in parallel without restarts.

**Comment by nroose** [comment]
For me, I don't get why we would want to have restarting pods and restarting containers in the pod. Especially if the pod has one container, why do we need to restart the container and why can't it just be a new pod. A deployment starts a new pod when the pod fails. Why do we need to implement all the features on all the different objects? Doesn't make sense to me. And somehow to prevent the never or on failure options it just because it does support always? That seems totally silly.

**Comment by Spenhouet** [comment]
> The pod consumes the node’s resources for having space for your containers. Killing the pod means losing this reservation.
> 
> I don’t think this is a good idea.

This is the whole point why we need this. We want the pod to lose the reservation. The pod blocks our limited GPU resources and we want them to be free for other pods. Having the deployments not exit out blocks our resources.

**Comment by bronger** [comment]
As pointed out earlier, then even a more permissive restartPolicy would not help you. You have to open a new issue.

**Comment by kyrylogy** [comment]
I can't use Job because of some other limitations, and Deployment does not allow for `restartPolicy: Never`. Absolutely stuck at this point

**Comment by hypdeb** [comment]
Is the documentation supposed to be up to date? I feel like I'm missing something here: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

**Comment by sanmai-NL** [comment]
@scarby 

A Deployment is useful over a single Pod, since Deployments can be updated with `kubectl apply` whereas the Pod would have to be deleted and recreated manually in some cases. This is advantageous when a container image (e.g., with a mutable tag `latest`) is being updated frequently (e.g., a development situation).

In that same debugging and development situation, you may want a LivenessProbe for the Pods managed by the Deployment, but may not want the Pods to be cycled automatically.

**Comment by JoeAshworth** [comment]
I genuinely cannot believe the amount of people requesting this and the stubbornness displayed in this issue. 

There is a clear use-case for this feature.

**Comment by dims** [comment]
@JoeAshworth https://twitter.com/dims/status/1593348306054397952?s=20&t=hK-e1SZ_nyaLwlGbYnN53g

**Comment by Typraeurion** [comment]
I think I have a different use case than the ones I’ve read above.  I have a sidecar which is failing on startup that I need to troubleshoot.  The kink is that it’s in an OpenShift environment where I don‘t have direct access via `kubectl`; I can only access the container logs via its web UI, and every time the container restarts the UI loses the old container’s log output.  I tried setting the restart policy to “Never” to prevent this from happening and see if that would preserve the logs long enough for me to read them, but got the error that this isn’t supported which led me here.

**Comment by pre** [comment]
> I think I have a different use case than the ones I’ve read above.

In this use case your cluster should stream the k8s logs to a long term storage (which you probably should do anyway). A common pattern is to eg run Fluentbit as a DaemonSet and stream the container logs from the k8s nodes to eg Elasticsearch. This allows indexing, searching and browsing the logs with eg Kibana. It also decouples the access control of log viewing from access control if the k8s resources (kubectl logs).

**Comment by nroose** [comment]
I mean is there some practical reason that the deployment can’t support
never or is it just a principle or something? Can we please have it?

**Comment by athornton** [comment]
Or if you won't fix it, could you please fix the documentation, specifically https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy ?

**Comment by RiccardoManzan** [comment]
I'd like to have the ability to set Never in order to be able to debug in a pod which is having some issues only in a kubernetes environment.

**Comment by woj-tek** [comment]
I just ran into same problem as @RiccardoManzan and quick search brought me to this exact issue...

**Comment by Stef16Robbe** [comment]
Same problem here! Still going...

**Comment by rivamarco** [comment]
Same issue here, I need to leave a pod inside a deployment in failed state without rebooting but it's impossible

**Comment by lwa144** [comment]
I have the same issue and came into the discussion, haha, what a long talk.

**Comment by kerthcet** [comment]
Better to provide some context here like why you need this capacity or what you have tried but not ideal for use case collections.

**Comment by rivamarco** [comment]
> Better to provide some context here like why you need this capacity or what you have tried but not ideal for use case collections.

I would like to be able to do not restart deployment in order to understand what happened if failed and block the constant reboot that may happen, also during debugging for example.

Now I'm trying to do a workaround changing the Deployment into a Job because in that way it seems possible to do not restart on failure

**Comment by jtdhyl** [comment]
pod always restart and i cant delete it，because of restartPolicy only support always，it is Very foolish

**Comment by jtdhyl** [comment]
So many years have passed and so many people propose this question，they just dont solve it ，i cant understand

**Comment by wvxvw** [comment]
I'll throw another rock into this huge pile.

So, I was working on upgrading Kubernetes (oh noes!) among mountains of things that don't work, there's this particular tidbit: after you've done running `kubadm upgrade apply`, you need to drain the node. And if you happen to have single pods on that node not managed by deployments / replica sets / daemon sets and so on -- well, tough luck, instead of moving your pods to a different node they are going to be killed.

Sucks, right? -- well, that's not the end of the story... I was about to get smart and just automatically wrap any pod in a deployment object -- how bad can it be, right? -- and now I'm here -- fun, fun, fun!

**Comment by 2PacIsAlive** [comment]
> I'll throw another rock into this huge pile.
> 
> So, I was working on upgrading Kubernetes (oh noes!) among mountains of things that don't work, there's this particular tidbit: after you've done running `kubadm upgrade apply`, you need to drain the node. And if you happen to have single pods on that node not managed by deployments / replica sets / daemon sets and so on -- well, tough luck, instead of moving your pods to a different node they are going to be killed.
> 
> Sucks, right? -- well, that's not the end of the story... I was about to get smart and just automatically wrap any pod in a deployment object -- how bad can it be, right? -- and now I'm here -- fun, fun, fun!

Is it possible to just wrap it in a Job instead of a Deployment?

**Comment by wvxvw** [comment]
Just a heads up. It kinda failed sooner than I thought. NVidia container runtime pods / deployments don't survive upgrade regardless what I wrap them in. So, at least for me, it doesn't matter if jobs worked better than deployments. I'm going to kill and bury those pods anyways.

**Comment by ryanovas** [comment]
Is there anywhere else this is still open and in progress? I would like to be able to set Never so if I have a failing pod in a deployment I can get in to debug it easier without it constantly restarting and losing logs/ability to exec in...

**Comment by OverStruck** [comment]
And the documentation hasn't been updated to let us know deployments only support "Always" for restartPolicy, thanks guys...

**Comment by remram44** [comment]
If you spotted a place where it's still missing, open an issue or a pull request please, instead of sending this vague passive-aggresive comment on a closed issue.

**Comment by balbert-etraid** [comment]
Any progress regarding this? Thanks in advance

**Comment by Sceat** [comment]
So I wanted to prevent my pods from restarting in order to find out the cause of the crash, I come accross this 8 years old issue and we still can't do that ? lmao this is wild

**Comment by nikhisin3001** [comment]
Yup still cant in 2025 :/

**Comment by laithalissa** [comment]
Just a heads up, I bumped my head on this after implementing something against my api server's docs:
```
❯ kubectl explain deployment.spec.template.spec.restartPolicy
GROUP:      apps
KIND:       Deployment
VERSION:    v1

FIELD: restartPolicy <string>
ENUM:
    Always
    Never
    OnFailure

DESCRIPTION:
    Restart policy for all containers within the pod. One of Always, OnFailure,
    Never. In some contexts, only a subset of those values may be permitted.
    Default to Always. More info:
    https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

    Possible enum values:
     - `"Always"`
     - `"Never"`
     - `"OnFailure"`
```

I guess this is serving me the pod template spec documentation in this case, still a bit misleading imo. 

seeing this on `Server Version: v1.31.6-gke.1020000`
